doc/reference/index.rst

```diff
@@ -23,3 +23,203 @@ API Reference
+
+The reschedule attribute is used on a function that can reach a
+:ref:`reschedule point <scheduling_v2>` within its execution.
+
+Details
+-------
+
+The significance of this attribute is that when a rescheduling function
+is invoked by a thread it is possible for that thread to be suspended as
+a consequence of a higher-priority thread being made ready.  Whether the
+suspension actually occurs depends on the operation associated with the
+reschedule point and the relative priorities of the invoking thread and
+the head of the ready queue.
+
+Note that in the case of timeslicing, or reschedule points executed from
+interrupts, any thread may be suspended in any function.
+
+Functions that are not **reschedule** may be invoked from either thread
+or interrupt context.
+
+Functions that are **reschedule** may be invoked from thread context.
+
+Functions that are **reschedule** but not **sleep** may be invoked from
+interrupt context.
+
+.. _api_term_sleep:
+
+sleep
+=====
+
+The sleep attribute is used on a function that can cause the invoking
+thread to sleep.
+
+Explanation
+-----------
+
+This attribute is of relevance specifically when considering
+applications that use only non-preemptible threads, because the kernel
+will not replace a running cooperative-only thread at a reschedule point
+unless that thread has explicitly invoked an operation that caused it to
+sleep.
+
+This attribute does not imply the function will sleep unconditionally,
+but that the operation may require an invoking thread that would have to
+suspend, wait, or invoke :cpp:func:`k_yield()` before it can complete
+its operation.  This behavior may be mediated by **no-wait**.
+
+Functions that are **sleep** are implicitly **reschedule**.
+
+Functions that are **sleep** may be invoked from thread context.
+
+Functions that are **sleep** may be invoked from interrupt and
+pre-kernel contexts if and only if invoked in **no-wait** mode.
+
+.. _api_term_no-wait:
+
+no-wait
+=======
+
+The no-wait attribute is used on a function to indicate that a parameter
+to the function can force an execution path that will not cause the
+invoking thread to sleep.
+
+Explanation
+-----------
+
+The paradigmatic case of a no-wait function is a function that takes a
+timeout, to which :c:macro:`K_NO_WAIT` can be passed.  The semantics of
+this special timeout value are to execute the function's operation as
+long as it can be completed immediately, and to return an error code
+rather than sleep if it cannot.
+
+It is use of the no-wait feature that allows functions like
+:cpp:func:`k_sem_take` to be invoked from ISRs, since it is not
+permitted to sleep in interrupt context.
+
+A function with a no-wait path does not imply that taking that path
+guarantees the function is synchronous.
+
+Functions with this attribute may be invoked from interrupt and
+pre-kernel contexts only when the parameter selects the no-wait path.
+
+Functions that are **no-wait** functions are implicitly **sleep**.
```

Comment:
Does it mean that `no-wait` attribute implies that kernel calls are used (the one that can take K_NO_WAIT)? If that's true then `no-wait` functions cannot be called from pre-kernel (assuming kernel api shouldn't be used in pre-kernel).

Reply:
I've removed that line.  Yes, you can use no-wait functions from pre-kernel as long as they're pre-kernel-ok (which is also the default)

Absent any behavior specification every function is **no-wait**.

If the function is **sleeps** then it can wait,  so if there's a way to tell it to not sleep it needs **no-wait** added explicitly.

The idea was that from an explicit **no-wait** annotation you can infer that the function must be **sleeps**, because otherwise there would have been no need to say it was no-wait.

This concept of hierarchical interpretation of attributes with default values that depend on the setting of other tags has proven too confusing.  We'll have to work around that either by adding all relevant tags explicitly in the header, or by carefully crafted tooling that does the inference correctly.